### PR TITLE
Support third-party devices to use the init_process_group method with…

### PR DESCRIPTION
**Pull request by QSHLGZ**
…out specifying the Backend

When init_process_group is not been done before, it will automatically apply  init_process_group within Devicemesh without specifying the backend. Thus, when a third-party device want to use Devicemesh without doing init_process_group before, there comes a problem. In this PR, add a default_device_backend_map for third-party device users to add their backends to this map when they register their backends to pytorch firstly. When doing init_process_group without parameter backend, it will init the backends in this map. Thus, a third-party user can use init_process_group method without specifying the Backend.

### CODE REVIEW

**Comment by colesbury** [comment]
@mrshenli would you please review this or route it to the right reviewer?

**Review by wanchaol — APPROVED** [review]
lgtm

**Comment by wanchaol** [comment]
cc @H-Huang @kwen2501 This is something we have discussed before, I think @QSHLGZ proposed a reasonable change, so I'm going to land this

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -234,6 +239,10 @@ def register_backend(cls, name, func, extended_api=False, devices: Optional[Unio
 
         setattr(Backend, name.upper(), name.lower())
         Backend.backend_list.append(name.lower())
+        if devices is not None:
+            for device in devices:
+                if device != 'cpu' and device != 'cuda':
```
Comment (wanchaol) [inline torch/distributed/distributed_c10d.py:244]: question (can be addressed in the next PR): why we skip for cpu and cuda, if user register a custom backend for cpu/cuda device, it is highly likely user want to use this backend for cpu/cuda, I think a default backend should also be switch to the custom registered backend?

Comment (QSHLGZ) [reply torch/distributed/distributed_c10d.py:244]: There are two reasons: 1.Not change the current default process when doing init_process_group without specifying backend. 2.If without the device-check, when the function 'register_backend' is called, some backends like 'fake', 'threaded' for cpu will also be added in the default_device_backend_map.
Maybe in the next pr, I can extract this fuction to a new interface. When users want to register a new backend to the default_device_backend_map,  they can call the new interface, which can perfectly solve the two problems above.

**Pull request closed without merging**